### PR TITLE
[NativeAOT] Objective-C exception propagation

### PR DESCRIPTION
**Pull request by AustinWise**
Related issue: #77472

This is the final part of the Objective-C Marshal API: the ability to turn a managed exception unwinding across a reverse-pinvoke boundary into a native exception.

The pieces to support this include:

* If the first pass of exception unwinding hits a reverse-P/Invoke boundary, call a new classlib function to get the native callback.
* After performing the second pass of unwinding, use a new `RhpCallManagedToNativeExceptionCallback` assembly helper to restore the context and jump to the native callback. This helper is based off of `RhpCallCatchFunclet`, with the calling of catch funclet and thread-abort handling removed.

### Issues that need to be resolved before merging

* [x] add AMD64 version of the assembly helper
* [x] Figure what to do about `ThreadAbortException`, which will be swallowed if it goes through this code path. Perhaps change `RhpPInvokeReturn` to check for pending thread aborts? Or maybe thread aborts don't matter since `RhpInitiateThreadAbort` is not called by anything?
* [x] Simplify the prolog of the ARM64 assembly helper, since it does not need to compatible with `StackFrameIterator::UnwindFuncletInvokeThunk`.
* [x] Debug why AMD64 crashes.
* [x] Unwind to the correct frame.
* [x] Fix Windows build
* [x] Make `StackFrameIterator` support unwinding into a least one native frame
* [x] Create a tracking issue for implement the `lastMethod` parameter and update comments to reference it: #80985
* [x] ~When an unhandled exception hits a reverse P/Invoke function on Windows, fix backtrace from `RaiseFailFastException`.~ Moved to #81010.
* [x] Switch back to preemptive mode before jumping to exception propagation function.

### Testing done

Built and tested with these commands for ARM64. Similar invocations for testing with X64.

```
./build.sh -ninja -s clr.aot+libs /p:TestNativeAot=true -c Release
./src/tests/build.sh -nativeaot -ninja Release -tree:Interop/ObjectiveC/ObjectiveCMarshalAPI
export CLRCustomTestLauncher=`pwd`/src/tests/Common/scripts/nativeaottest.sh
export CORE_ROOT=`pwd`/artifacts/tests/coreclr/OSX.arm64.Release/Tests/Core_Root
./artifacts/tests/coreclr/OSX.arm64.Release/Interop/ObjectiveC/ObjectiveCMarshalAPI/ObjectiveCMarshalAPI.sh
```

The test passes, which means the C++ exception dispatch code is able to unwind the stack after the new assembly helper restored the context. I added some extra asserts to check that the second pass unwind is done and that a value in the frame being restored is not clobbered after execution resumes. Specifically the `assert(e == a)` assert loads `a` from the stack in debug and the caller-saved `x19` register in release. After jumping from the assembly helper to the native method that throws, the stack trace makes sense in the debugger (looks like `CallAndCatch` called `ThrowInt` or `ThrowException`).

I also tested this program continues to fail fast as expected on Windows:

```c#
using System;
using System.Runtime.InteropServices;

static class Program
{
    [UnmanagedCallersOnly]
    static void MyUnmanagedCallersOnlyFunction()
    {
        throw new Exception();
    }

    unsafe static void Main()
    {
        delegate* unmanaged<void> del = &MyUnmanagedCallersOnlyFunction;
        del();
    }
}
```

### Known limitations

The prolog of the AMD64 assembly helper could  preserve fewer registers.

The callback for exception unwinding expects a `RuntimeMethodHandle` that represents the managed method that was called via reverse P/Invoke. I tried doing this to convert the instruction pointer into a `RuntimeMethodHandle:

```c#
MethodBase? method = new StackFrame(ip, needFileInfo: false).GetMethod();
RuntimeMethodInfo? methodInfo = method as RuntimeMethodInfo;
RuntimeMethodHandle lastMethod = methodInfo?.MethodHandle ?? default;
```

For all the cases in the unit test, the `StackFrame` is unable to find the method object. I have not looked at this closely, but I believe this is because unmanaged-callers-only function and reverse-p/invoke stubs do not show up in the reflection invoke map. [See this part of ILC for some possibly related UCO logic](https://github.com/AustinWise/runtime/blob/889eb1949a17288c4d805486c6692c04cfe4aa63/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectionInvokeMapNode.cs#L177-L179).

Since the code to lookup the `RuntimeMethodHandle` didn't work and Xamarin [does not even use](https://github.com/xamarin/xamarin-macios/blob/6e1ba31bae568ec54f53259e9e884f0550cfdae3/src/ObjCRuntime/Runtime.CoreCLR.cs#L144) this value, I decided to remove the code and always set the `RuntimeMethodHandle` to 0.

### CODE REVIEW

**Comment by AustinWise** [comment]
@jankots
Does this general approach seem reasonable? If so, I'll add an AMD64 version of the assembly helper.

**Comment by MichalStrehovsky** [comment]
> The callback for exception unwinding expects a `RuntimeMethodHandle` that represents the managed method that was called via reverse P/Invoke. I tried doing this to convert the instruction pointer into a `RuntimeMethodHandle:

It is not guaranteed that we'll be able to produce a RuntimeMethodHandle for an arbitrary function pointer.

One thing is that they need to go to the mapping table, which currently skips them. That should not be hard to fix (we just need to do the right thing when attempting to reflection-invoke this, which is why it's blocked - the right thing being "whatever CoreCLR does").

The other thing is that the compiler will only generate these mapping table entries for methods that are visible targets of reflection. We don't have the names/signatures of methods that are not visible targets of reflection. If this is only needed for `UnmanagedCallersOnly` methods, you could make them visible targets of reflection by making UnmanagedCallersOnly on apple platforms take codepath around here, I think:

https://github.com/dotnet/runtime/blob/5da4a9e919dcee35f831ab69b6e475baaf798875/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs#L600-L601

It comes with size costs. It would be preferable if we don't need it.

File: src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs [hunk 1]
```diff
@@ -609,8 +609,11 @@ private static void DispatchEx(scoped ref StackFrameIterator frameIter, ref ExIn
 
             byte* prevControlPC = null;
             byte* prevOriginalPC = null;
+            UIntPtr prevControlSP = MaxSP;
             UIntPtr prevFramePtr = UIntPtr.Zero;
             bool unwoundReversePInvoke = false;
+            IntPtr pReversePinvokePropagationCallback = IntPtr.Zero;
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:615]:
```suggestion
            IntPtr pReversePInvokePropagationCallback = IntPtr.Zero;
```
Nit: The previous line uses `PInvoke` casing. It would be nice to be consistent.

**Comment by jkotas** [comment]
> Does this general approach seem reasonable? 

Yes, the approach looks reasonable.

> Figure what to do about ThreadAbortException

You do not need to worry about ThreadAbortException. It is not supported.

> The callback for exception unwinding expects a RuntimeMethodHandle

Xamarion iOS callback does not look at the RuntimeMethodHandle today: https://github.com/xamarin/xamarin-macios/blob/6e1ba31bae568ec54f53259e9e884f0550cfdae3/src/ObjCRuntime/Runtime.CoreCLR.cs#L144-L149 . It converts exceptions to for all reverse PInvokes even the ones that have nothing to do with ObjectiveC interop (it is less than ideal). It should be ok to pass in `default(RuntimeMethodHandle)` here unless Xamarin iOS implementation changes.

**Comment by jkotas** [comment]
> It should be ok to pass in default(RuntimeMethodHandle) here unless Xamarin iOS implementation changes.

We may want to consider introducing declarative scheme for this (mark all methods that want to participate in ObjectiveC interop with an attribute) to make this more pay-for-play and AOT friendly.

**Comment by AustinWise** [comment]
This is still not ready for review.

I simplified the ARM64 assembly helper and added an initial version of the AMD64 version. The AMD64 version is not yet working.

I took a more careful look at the stack pointers being restored, using some printf hacks. It appears that both ARM64 and x64 are restoring the stack pointer register to values from the managed function, not the calling `CallAndCatch` function as I intended. It appears that ARM64 is only working because the native exception unwinder is restoring the correct value before executing the catch block. The x64 version is not so lucky.

**Comment by AustinWise** [comment]
The test now passes on X64 too, after applying the same 8 byte offset to RSP that CoreCLR does:

https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/exceptionhandling.cpp#L4343-L4346

**Comment by AustinWise** [comment]
One thing I'm not sure is right is the back trace looks like this after jumping to [`ThrowInt` in the test](https://github.com/AustinWise/runtime/blob/8758680d7b93926085c5a4c434459fd63eaead14/src/tests/Interop/ObjectiveC/ObjectiveCMarshalAPI/NativeObjCMarshalTests.cpp#L101-L106):

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = step in
  * frame #0: 0x0000000102018a6c libNativeObjCMarshalTests.dylib`(anonymous namespace)::ThrowInt(cxt=-121) at NativeObjCMarshalTests.cpp:104:32
    frame #1: 0x00000001002f98ac ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program__UCO_ThrowIntException + 76
    frame #2: 0x0000000102018b08 libNativeObjCMarshalTests.dylib`::CallAndCatch(fptr=(ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program__UCO_ThrowIntException), a=3423) at NativeObjCMarshalTests.cpp:131:9
    frame #3: 0x00000001002f903a ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_NativeObjCMarshalTests__CallAndCatch + 106
    frame #4: 0x00000001002f9d97 ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program___Validate_ExceptionPropagation + 551
    frame #5: 0x00000001002f9842 ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program__Validate_ReferenceTracking_Scenario + 882
    frame #6: 0x00000001002f9ef5 ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program__Main + 53
    frame #7: 0x00000001003ad1ff ObjectiveCMarshalAPI`ObjectiveCMarshalAPI__Module___MainMethodWrapper + 15
    frame #8: 0x00000001003ad27c ObjectiveCMarshalAPI`ObjectiveCMarshalAPI__Module___StartupCodeMain + 108
    frame #9: 0x00000001000073ef ObjectiveCMarshalAPI`main(argc=1, argv=0x00000003059f3788) at main.cpp:218:12
    frame #10: 0x0000000201902310 dyld`start + 2432
```

Notice the managed function `UCO_ThrowIntException` still show up in the stack trace after jumping to the propagation function. Compare to CoreCLR, where the managed function does not show up in the backtrace:

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
  * frame #0: 0x000000010043fb94 libNativeObjCMarshalTests.dylib`(anonymous namespace)::ThrowInt(cxt=0x0000000000000d5f) at NativeObjCMarshalTests.cpp:105:9 [opt]
    frame #1: 0x000000010043fc0c libNativeObjCMarshalTests.dylib`::CallAndCatch(fptr=<unavailable>, a=3423) at NativeObjCMarshalTests.cpp:131:9 [opt]
    frame #2: 0x0000000118fbce1c
    frame #3: 0x0000000118fb7814
    frame #4: 0x0000000118fb0d98
    frame #5: 0x0000000102908f34 libcoreclr.dylib`CallDescrWorkerInternal + 132
    frame #6: 0x0000000102785284 libcoreclr.dylib`MethodDescCallSite::CallTargetWorker(unsigned long long const*, unsigned long long*, int) + 856
    frame #7: 0x000000010268bf4c libcoreclr.dylib`RunMain(MethodDesc*, short, int*, PtrArray**) + 656
    frame #8: 0x000000010268c228 libcoreclr.dylib`Assembly::ExecuteMainMethod(PtrArray**, int) + 396
    frame #9: 0x00000001026b3f94 libcoreclr.dylib`CorHost2::ExecuteAssembly(unsigned int, char16_t const*, int, char16_t const**, unsigned int*) + 616
    frame #10: 0x0000000102a3049c libcoreclr.dylib`coreclr_execute_assembly + 192
    frame #11: 0x0000000100005780 corerun`run(config=0x000000016fdff270) at corerun.cpp:429:18 [opt]
    frame #12: 0x0000000100002a88 corerun`main(argc=<unavailable>, argv=<unavailable>) at corerun.cpp:624:21 [opt]
    frame #13: 0x00000001a829be50 dyld`start + 2544
```

**Comment by jkotas** [comment]
> One thing I'm not sure is right is the back trace looks like this after jumping to [ThrowInt in the test](https://github.com/AustinWise/runtime/blob/8758680d7b93926085c5a4c434459fd63eaead14/src/tests/Interop/ObjectiveC/ObjectiveCMarshalAPI/NativeObjCMarshalTests.cpp#L101-L106):

Yes, it does not look right. All managed frames should be gone when the propagation function is called.

**Comment by AustinWise** [comment]
When I tried to step the StackFrameIterator past the reverse P/Invoke frame, I discovered that it skips directly from the reverse P/Invoke frame to the P/Invoke frame. This skips all the unmanaged frames in between:

https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp#L293-L318

https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime/StackFrameIterator.cpp#L1450-L1458

So I added some hacks to allow the the `StackFrameIterator` to unwind into unmanaged code. The backtrace from the `ThrowInt` no longer shows the managed `UCO_ThrowIntException` function:

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = step in
  * frame #0: 0x00000001019679e0 libNativeObjCMarshalTests.dylib`(anonymous namespace)::ThrowInt(cxt=0x0000000000000d5f) at NativeObjCMarshalTests.cpp:105:9
    frame #1: 0x0000000101967a60 libNativeObjCMarshalTests.dylib`::CallAndCatch(fptr=(ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program__UCO_ThrowIntException), a=3423) at NativeObjCMarshalTests.cpp:131:9
    frame #2: 0x00000001002d7d60 ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_NativeObjCMarshalTests__CallAndCatch + 112
    frame #3: 0x00000001002d8bd8 ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program___Validate_ExceptionPropagation + 536
    frame #4: 0x00000001002d860c ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program__Validate_ReferenceTracking_Scenario + 876
    frame #5: 0x00000001002d8d90 ObjectiveCMarshalAPI`ObjectiveCMarshalAPI_ObjectiveCMarshalAPI_Program__Main + 48
    frame #6: 0x00000001003d344c ObjectiveCMarshalAPI`ObjectiveCMarshalAPI__Module___MainMethodWrapper + 12
    frame #7: 0x00000001003d34c0 ObjectiveCMarshalAPI`ObjectiveCMarshalAPI__Module___StartupCodeMain + 96
    frame #8: 0x000000010000cb50 ObjectiveCMarshalAPI`main(argc=1, argv=0x000000016fdff688) at main.cpp:218:12
    frame #9: 0x00000001a829be50 dyld`start + 2544
```

@jkotas What do you think of the idea enabling the `StackFrameIterator` to unwind into unmanaged code? It does not look like it was really designed with this in mind, so any ideas on how to more elegantly enable this would be appreciated.

**Comment by jkotas** [comment]
> @jkotas What do you think of the idea enabling the StackFrameIterator to unwind into unmanaged code? It does not look like it was really designed with this in mind, so any ideas on how to more elegantly enable this would be appreciated.

- Add a new `StackFrameIterator::Flags` that says whether to skips directly from the reverse P/Invoke frame to the P/Invoke frame. This flag should be set for GC stack walks, it should be cleared for EH stack walks. It will avoid need for the hacky `IsEH` method.
- Add a new argument to the existing `UnwindStackFrame` method. I think it would look better than introducing a whole new method.

**Comment by AustinWise** [comment]
This is now ready for review.

I noticed what appears to be a bug for Windows. When calling `RaiseFailFastException` on an unhanded exception, the `CONTEXT` parameter contains uninitialized stack memory. Depending on what bytes happen to be on the stack, the call stack displayed in Visual Studio after calling `RaiseFailFastException` is corrupted. Zero-initializing the `CONTEXT` appears to fix the problem. Perhaps this fix is worth backporting to .NET 7, since it is small in scope and could aid in diagnosing crashes.

**Review by jkotas — APPROVED** [review]
LGTM! @VSadov @janvorli @MichalStrehovsky Any additional feedback?

File: src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S [hunk 2]
```diff
@@ -532,3 +532,146 @@ NESTED_ENTRY RhpCallFilterFunclet, _TEXT, NoHandler
...
+//
+// INPUT:  RDI:  callback context
+//         RSI:  callback
+//         RDX:  REGDISPLAY*
+//         RCX:  ExInfo*
+//
+// OUTPUT:
+//
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////
+NESTED_ENTRY RhpCallPropagateExceptionCallback, _TEXT, NoHandler
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:552]: Where does the switch to preemptive mode happens as we transition out of the managed code? Do we need to switch to preemptive mode somewhere in this method?

File: src/coreclr/nativeaot/Runtime/EHHelpers.cpp [hunk 3]
```diff
@@ -91,8 +91,8 @@ COOP_PINVOKE_HELPER(int32_t, RhGetModuleFileName, (HANDLE moduleHandle, _Out_ co
 
 COOP_PINVOKE_HELPER(void, RhpCopyContextFromExInfo, (void * pOSContext, int32_t cbOSContext, PAL_LIMITED_CONTEXT * pPalContext))
 {
-    UNREFERENCED_PARAMETER(cbOSContext);
     ASSERT((size_t)cbOSContext >= sizeof(CONTEXT));
+    memset(pOSContext, 0, cbOSContext);
     CONTEXT* pContext = (CONTEXT *)pOSContext;
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/EHHelpers.cpp:96]: I am wondering whether we need to set `ContextFlags` here as well to indicate which registers are valid.

File: src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S (hunk 2 above)
Comment (AustinWise) [reply src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:552]: I believe the thread state flag `TSF_DoNotTriggerGc` corresponds to cooperative mode, correct? It is cleared before restoring the registers:

```asm
// Clear the DoNotTriggerGc state before calling out to the propagation callback.
lock and     dword ptr [rax + OFFSETOF__Thread__m_ThreadStateFlags], ~TSF_DoNotTriggerGc
```

https://github.com/dotnet/runtime/pull/80334/files/d2b5cd8a1807577f4895e44f77eb4fa1a3ea17f4#diff-6d87d47bde66f3bf88ea6a7e87f5ceba784e8f02024ef1b7beda7a6ad7ba50a6R575-R576

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:552]: `TSF_DoNotTriggerGc` does not indicate preemptive/cooperative mode.

The thread is in preemptive mode when `Thread::m_pTransitionFrame` is non-null.

You can check for problems by calling `[UnmanagedCallersOnly]` callback from the unmanaged propagation function. If the mode is not switched properly, you will hit `Invalid Program: attempted to call a UnmanagedCallersOnly method from managed code` failfast.

File: src/coreclr/nativeaot/Runtime/EHHelpers.cpp (hunk 3 above)
Comment (AustinWise) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp:96]: That's a good point. I experimented with setting the `ContextFlags` to `CONTEXT_CONTROL | CONTEXT_INTEGER` and setting the correct values for the `CS` and `SS` segment registers. Once I did that, the call stack in Visual Studio after calling `RaiseFailFastException` directly pointed to the function that threw the unhandled exception.

Since this change is unrelated to this PR and will require a little more work to capture these segment registers, I'm going to move these change to a separate PR.

Comment (AustinWise) [reply src/coreclr/nativeaot/Runtime/EHHelpers.cpp]: I'll continue the work here: #81010.

File: src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S (hunk 2 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:552]: `TSF_DoNotTriggerGc` makes a thread uninterruptible/unhijackable when it runs managed code. I.E. managed code called form GC should not initiate or try to stop for GC as that would deadlock, so GC callouts set this flag.

The use of  `TSF_DoNotTriggerGc` in exception handling is less clear. I found few comments on it and they are a bit confusing  - 
https://github.com/dotnet/runtime/blob/b2dc37ba181a7fa4427e717eab819ba3543d0ae4/src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.asm#L543

The idea sems to be to disallow hijacking once the stack is inconsistent state due to exception unwind. Indeed we could have a hijack in the portion of the stack that just went dead.  
It looks like we permit hijacking in funclets, but not "between" funclets.

The flag certainly does not switch the thread to preemptive mode.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:552]: More relevant comment - 
https://github.com/dotnet/runtime/blob/8918adf2e5255920aed3c61f9b92c2a51eec0fe3/src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs#L672

We do not allow GC in managed EH dispatch code once we are in the second pass. 
GC in funclets is ok though, thus it is toggled when calling the funclets.

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:552]: If I understand this correctly - we are done with exception dispatching and are basically passing the control to a native handler while the managed frames that were reverse-pinvoked and faulted are effectively gone. 
If GC happens, it will need to scan the current stack starting from some higher managed frame, if there was one, that called the native code, that called the managed code which faulted.

Basically we need to fetch the innermost pinvoke frame (`m_savedPInvokeTransitionFrame`) and put it in the `m_pTransitionFrame` - the same way as return from reverse pinvoke does. It could be `TOP_OF_STACK_MARKER` if there are no more managed frames.

`TSF_DoNotTriggerGc` should be cleared too, since we are done dispatching.

Comment (AustinWise) [reply src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:552]: Ah, I think I see what you are saying. Since this unwinds pass the reverse P/Invoke frame, it never invokes `RhpReversePInvokeReturn`, so the thread won't call this to return to preemptive mode:

https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime/thread.cpp#L1234-L1238

The `StackFrameIterator` does find the `PInvokeTransitionFrame` when unwinding, so perhaps I could restore that that to re-enable preemptive mode?

**Comment by MichalStrehovsky** [comment]
> LGTM! @VSadov @janvorli @MichalStrehovsky Any additional feedback?

Not my area of expertise, but looks good to me!

File: src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S (hunk 2 above)
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:552]: Yes, that should work.

File: src/tests/Interop/ObjectiveC/ObjectiveCMarshalAPI/Program.cs [hunk 4]
```diff
@@ -282,6 +289,13 @@ static unsafe void Validate_ReferenceTracking_Scenario()
             _Validate_ExceptionPropagation();
         }
 
+        [UnmanagedCallersOnly]
+        private static void BeforeThrowNativeException()
+        {
+            // This function is called from the exception propagation callback.
+            // It ensures that the thread was transitioned to preemptive mode.
```
Comment (jkotas) [inline src/tests/Interop/ObjectiveC/ObjectiveCMarshalAPI/Program.cs:296]:
```suggestion
            // It ensures that the thread was transitioned to preemptive mode.
            GC.Collect();
```
Triggering GC verifies that the runtime and thread stack are in a good state.

**Comment by VSadov** [comment]
asm offsets on windows look incorrect

**Comment by AustinWise** [comment]
> asm offsets on windows look incorrect

Fixed; too bad that MSVC does not print the expression in `static_assert`.

File: src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S (hunk 2 above)
Comment (AustinWise) [reply src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.S:553]: Done. I re-ran the tests on arm64 and x64 (well, Rosetta 2 technically) to confirm that the augmented test now passes.

**Review by VSadov — APPROVED** [review]
LGTM Thanks!

**Comment by VSadov** [comment]
The failures in extra-platforms do not look related to this change.

```
System.MissingMethodException : Constructor on type 'Microsoft.Extensions.Logging.LoggerFactoryOptions' not found.
   at System.Activator.CreateInstance[T]() + 0x94
   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String) + 0x38
   at Microsoft.Extensions.Options.UnnamedOptionsManager`1.get_Value() + 0xb4
   at Microsoft.Extensions.Logging.LoggerFactory..ctor(IEnumerable`1 providers, IOptionsMonitor`1 filterOption, IOptions`1 options, IExternalScopeProvider scopeProvider) + 0xcc
   at Microsoft.Extensions.Logging!<BaseAddress>+0x78fce8
   at System.Reflection.DynamicInvokeInfo.Invoke(Object, IntPtr, Object[], BinderBundle, Boolean) + 0x14c
   at Internal.Reflection.Execution.MethodInvokers.InstanceMethodInvoker.Invoke(Object, Object[], BinderBundle, Boolean) + 0x60
   at Internal.Reflection.Core.Execution.MethodInvoker.Invoke(Object, Object[], Binder, BindingFlags, CultureInfo) + 0x54
   at System.Reflection.Runtime.MethodInfos.RuntimePlainConstructorInfo`1.Invoke(BindingFlags, Binder, Object[], CultureInfo) + 0xa8
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite, RuntimeResolverContext) + 0x84
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument) + 0xe0
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite, ServiceProviderEngineScope) + 0x34
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(Type serviceType) + 0xec
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey, Func`2) + 0x9c
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type, ServiceProviderEngineScope) + 0x3c
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider, Type) + 0x64
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider) + 0x30
   at Microsoft.Extensions.Logging.Test.LoggerTest.Log_IgnoresExceptionInIntermediateLoggersAndThrowsAggregateException() + 0x70
   at Microsoft.Extensions.Logging!<BaseAddress>+0x73d730
   at System.Reflection.DynamicInvokeInfo.Invoke(Object, IntPtr, Object[], BinderBundle, Boolean) + 0xd8
```

**Comment by MichalStrehovsky** [comment]
> The failures in extra-platforms do not look related to this change.
> 
> ```
> System.MissingMethodException : Constructor on type 'Microsoft.Extensions.Logging.LoggerFactoryOptions' not found.
>    at System.Activator.CreateInstance[T]() + 0x94
>    at Microsoft.Extensions.Options.OptionsFactory`1.Create(String) + 0x38
>    at Microsoft.Extensions.Options.UnnamedOptionsManager`1.get_Value() + 0xb4
>    at Microsoft.Extensions.Logging.LoggerFactory..ctor(IEnumerable`1 providers, IOptionsMonitor`1 filterOption, IOptions`1 options, IExternalScopeProvider scopeProvider) + 0xcc
>    at Microsoft.Extensions.Logging!<BaseAddress>+0x78fce8
>    at System.Reflection.DynamicInvokeInfo.Invoke(Object, IntPtr, Object[], BinderBundle, Boolean) + 0x14c
>    at Internal.Reflection.Execution.MethodInvokers.InstanceMethodInvoker.Invoke(Object, Object[], BinderBundle, Boolean) + 0x60
>    at Internal.Reflection.Core.Execution.MethodInvoker.Invoke(Object, Object[], Binder, BindingFlags, CultureInfo) + 0x54
>    at System.Reflection.Runtime.MethodInfos.RuntimePlainConstructorInfo`1.Invoke(BindingFlags, Binder, Object[], CultureInfo) + 0xa8
>    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite, RuntimeResolverContext) + 0x84
>    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument) + 0xe0
>    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite, ServiceProviderEngineScope) + 0x34
>    at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(Type serviceType) + 0xec
>    at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey, Func`2) + 0x9c
>    at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type, ServiceProviderEngineScope) + 0x3c
>    at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider, Type) + 0x64
>    at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider) + 0x30
>    at Microsoft.Extensions.Logging.Test.LoggerTest.Log_IgnoresExceptionInIntermediateLoggersAndThrowsAggregateException() + 0x70
>    at Microsoft.Extensions.Logging!<BaseAddress>+0x73d730
>    at System.Reflection.DynamicInvokeInfo.Invoke(Object, IntPtr, Object[], BinderBundle, Boolean) + 0xd8
> ```

I think it's from @vitek-karas change. I submitted a revert of his pr and it doesn't show these failures: #81259

**Comment by MichalStrehovsky** [comment]
This change broke something around unhandled exceptions on Windows.

With the Convenience Visual Studio repro project workflow from here: https://github.com/dotnet/runtime/blob/f1bdd5a6182f43f3928b389b03f7bc26f826c8bc/docs/workflow/building/coreclr/nativeaot.md#convenience-visual-studio-repro-project. Place a `throw new Exception()` in main.

Previously, the experience looked like this in the debugger:

image


Now it looks like this (note "Frame not in module" and a useless stack):

image

This change made it into the Preview 1 snap. Unless we roll back, unhandled exceptions are going to be undiagnosable in Preview 1, at least on Windows.

**Comment by AustinWise** [comment]
> This change broke something around unhandled exceptions on Windows.
> 
> With the Convenience Visual Studio repro project workflow from here: https://github.com/dotnet/runtime/blob/f1bdd5a6182f43f3928b389b03f7bc26f826c8bc/docs/workflow/building/coreclr/nativeaot.md#convenience-visual-studio-repro-project. Place a `throw new Exception()` in main.
> 
> Previously, the experience looked like this in the debugger:
> 
> image
> 
> Now it looks like this (note "Frame not in module" and a useless stack):
> 
> image
> 
> This change made it into the Preview 1 snap. Unless we roll back, unhandled exceptions are going to be undiagnosable in Preview 1, at least on Windows.

This is should be fixed by #81010 .

**Comment by AustinWise** [comment]
@MichalStrehovsky Just to be clear, this problem was reproducible before this PR was merged. This PR incidentally has groomed the stack in a way that makes `RaiseFailFastException` unhappy when it reads the uninitialized memory.

#81010 should probably be considered for servicing for 7.0 as well.

**Comment by MichalStrehovsky** [comment]
My only gripe is that this is the first time I'm observing this failure mode and we just snapped 8.0 Preview 1 a couple hours ago, which means we're going to ship preview 1 like this. It's not the end of the world, but not great either.

Cc @agocke - see https://github.com/dotnet/runtime/pull/80334#issuecomment-1409748951

**Comment by AustinWise** [comment]
Sorry about that. In retrospect I should have made sure the PR fixing the uninitialized memory read landed before this PR made the call stacks useless. I'll keep that principal in mind going forward.
